### DISCUSSION TITLE
How can I determine if a ListItem field already has a value?

### DISCUSSION THREAD

**Discussion by adamfisher** [pnp/pnpcore#1229] [Q&A] [answered]
This sounds simple and should be easy but for whatever reason SharePoint is a pain and any help would be appreciated. I'm trying to populate the `hasProjectNumber` boolean variable with true if that field already has a value in that SP column. **I verified in the UI that there is a value for the folder I am currently debugging against.** I just can't seem to get the existing value in that field?

```chsarp
foreach(IChange change in changes) {
  if (change is IChangeItem changeItem &&
    changeItem.IsPropertyAvailable < IChangeItem > (p => p.ListId) &&
    changeItem.IsPropertyAvailable < IChangeItem > (p => p.ItemId)) {

    IList ? list = await _sharePointContext.Web.Lists.GetByIdAsync(changeItem.ListId);

    IListItem item = await list.Items.GetByIdAsync(changeItem.ItemId,
      li => li.FieldValuesForEdit,
      li => li.File.QueryProperties(f => f.Name, f => f.ServerRelativeUrl),
      li => li.Folder.QueryProperties(f => f.Name, f => f.ServerRelativeUrl));

    bool alreadyHasProjectId = await item.FieldIsNotEmptyAsync("ProjectNumber");
	
	var hasProjectNumber = item.Values.ContainsKey("ProjectNumber") && !string.IsNullOrWhiteSpace(item["ProjectNumber"] as string);
	
	// hasProjectNumber is always false...
  }
}
```

Maybe it has something to do with how I am loading the item? I've tried setting up my query to [include fields like in the doc example](https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#a-getting-list-items-max-100-items) but that didn't change anything. I also should mention **later in my code I am able to assign to that field and see the value reflected in SP UI** so I know it is a valid field. I just can't get the existing value.

**Comment by adamfisher** [comment]
I've tried these variations too but they also produce errors or false:

```csharp
// KeyNotFoundException
var hasProjectNumber = item["ProjectNumber"] != null;

// Always false
var hasProjectNumber = item.File.ListItemAllFields.Values.ContainsKey("ProjectNumber")
   || item.Folder.ListItemAllFields.Values.ContainsKey("ProjectNumber");
```

**Reply by jansenbe** [reply]
@adamfisher : the latter approach should work fine, we use it all the time in internal code. E.g. see below snippet from the Pages API code:

```csharp
if (item.Values.ContainsKey(PageConstants.ClientSideApplicationId) && 
    item.Values[PageConstants.ClientSideApplicationId] != null && 
    item.Values[PageConstants.ClientSideApplicationId].ToString().Equals(PageConstants.SitePagesFeatureId, StringComparison.InvariantCultureIgnoreCase))
{
}
```

Think the issue is with how you load the IListItem, seems the fields you need were not loaded.

**Reply by adamfisher** [reply, marked as answer]
Thanks Bert but it looks like the only way I was able to get the properties to load was by using the **All** property when loading the item:

```csharp
IListItem item = await list.Items.GetByIdAsync(changeItem.ItemId,
      li => li.All,
      li => li.File.QueryProperties(f => f.Name, f => f.ServerRelativeUrl),
      li => li.Folder.QueryProperties(f => f.Name, f => f.ServerRelativeUrl));
```

It didn't load it with any other combination of properties I tried to load. Not sure if that is a bug or just the way the API works but feels hacky. I also tried `li => li.Values` but that produces an error.
